src/Zed/Persistence/Repository/RepositoryPropelQueryJoinRule.php

```diff
@@ -0,0 +1,254 @@
+            $this->applyRule($methodTransfer, $tableTransferCollection, $classNodeTransfer);
+        }
+    }
+
+    /**
+     * @param \ArchitectureSniffer\PropelQuery\Method\Transfer\MethodTransfer $methodTransfer
+     * @param \ArchitectureSniffer\PropelQuery\Schema\Transfer\PropelSchemaTableTransfer[] $tableTransferCollection
+     * @param \ArchitectureSniffer\PropelQuery\ClassNode\Transfer\ClassNodeTransfer $classNodeTransfer
+     *
+     * @return void
+     */
+    protected function applyRule(
+        MethodTransfer $methodTransfer,
+        array $tableTransferCollection,
+        ClassNodeTransfer $classNodeTransfer
+    ): void {
+        $relationNames = $methodTransfer->getRelationNames();
+
+        $moduleNames = [];
+        $declaredModules = $methodTransfer->getDeclaredDependentModuleNames();
+
+        foreach ($relationNames as $relationTableName) {
+            $relationTableTransfer = $this->findRelationTableTransfer(
+                $relationTableName,
+                $tableTransferCollection
+            );
+
+            if ($relationTableTransfer === null) {
+                continue; //todo: must be deleted with implementation AddJoinCondition and Add JoinObject
+            }
+
+            $moduleName = $relationTableTransfer->getModuleName();
+
+            $ownerModuleName = $classNodeTransfer->getClassModuleName();
+
+            if ($moduleName === $ownerModuleName) {
+                continue;
+            }
+
+            $moduleNames[] = $moduleName;
+
+            if (in_array($moduleName, $declaredModules)) {
+                continue;
+            }
+
+            $this->addViolationMessage(
+                $methodTransfer->getMethodName(),
+                $relationTableName,
+                $moduleName,
+                $classNodeTransfer->getNode()
+            );
+        }
+
+        $extraDeclaredModules = array_diff($declaredModules, $moduleNames);
+
+        if ($extraDeclaredModules !== []) {
+            $this->addExtraDeclaredModulesViolationMessage(
+                $methodTransfer->getMethodName(),
+                $extraDeclaredModules,
+                $classNodeTransfer->getNode()
+            );
+        }
+    }
+
+    /**
+     * @return \ArchitectureSniffer\PropelQuery\PropelQueryFacadeInterface
+     */
+    protected function getFacade(): PropelQueryFacadeInterface
+    {
+        if (!$this->facade) {
+            $this->facade = new PropelQueryFacade();
+        }
+
+        return $this->facade;
+    }
+
+    /**
+     * @param string $methodName
+     * @param string $relationTableName
+     * @param string $moduleName
+     * @param \PHPMD\Node\ClassNode $node
+     *
+     * @return void
+     */
+    protected function addViolationMessage(string $methodName, string $relationTableName, string $moduleName, ClassNode $node): void
+    {
+        $message = sprintf(
+            'The Repository method %s() violates rule "%s". ',
```

Comment:
Please make sure you use the same code fencing everywhere. @dereuromark prefers ` (backticks)